File: Frontend/samfms/src/backend/api/assignments.js
/**
 * Vehicle Assignment Management API
 * All vehicle assignment-related API endpoints and functions
 * Enhanced integration with Management Service and standardized error handling
 */
import { httpClient } from '../services/httpClient';
import { API_ENDPOINTS, buildApiUrl } from '../../config/apiConfig';
import {
  withRetry,
  handleApiResponse,
  parseApiError,
  validateRequiredFields,
  ERROR_TYPES,
} from '../../utils/errorHandler';

// Vehicle Assignment API endpoints using centralized configuration
const ASSIGNMENT_ENDPOINTS = {
  list: API_ENDPOINTS.ASSIGNMENTS.LIST,
  create: API_ENDPOINTS.ASSIGNMENTS.CREATE,
  update: API_ENDPOINTS.ASSIGNMENTS.UPDATE,
  delete: API_ENDPOINTS.ASSIGNMENTS.DELETE,
  metrics: buildApiUrl('/analytics/assignment-metrics'),
  complete: API_ENDPOINTS.ASSIGNMENTS.COMPLETE,
  cancel: API_ENDPOINTS.ASSIGNMENTS.CANCEL,
};

/**
 * Get list of vehicle assignments with optional filters
 * @param {Object} params - Query parameters (skip, limit, vehicle_id, driver_id)
 * @returns {Promise<Object>} Assignments list with pagination info
 */
export const getVehicleAssignments = async (params = {}) => {
  return withRetry(async () => {
    const queryParams = new URLSearchParams();
    if (params.skip) queryParams.append('skip', params.skip);
    if (params.limit) queryParams.append('limit', params.limit);
    if (params.vehicle_id) queryParams.append('vehicle_id', params.vehicle_id);
    if (params.driver_id) queryParams.append('driver_id', params.driver_id);

    const endpoint = `${ASSIGNMENT_ENDPOINTS.list}${
      queryParams.toString() ? '?' + queryParams.toString() : ''
    }`;

    const response = await httpClient.get(endpoint);
    return handleApiResponse(response);
  });
};

/**
 * Create a new vehicle assignment
 * @param {Object} assignmentData - Assignment data to create
 * @returns {Promise<Object>} Created assignment data
 */
export const createVehicleAssignment = async assignmentData => {
  validateRequiredFields(assignmentData, ['vehicle_id', 'driver_id']);

  return withRetry(
    async () => {
      const response = await httpClient.post(ASSIGNMENT_ENDPOINTS.create, assignmentData);
      return handleApiResponse(response);
    },
    { maxRetries: 1 }
  );
};

/**
 * Update a vehicle assignment
 * @param {string} assignmentId - Assignment ID
 * @param {Object} updateData - Data to update
 * @returns {Promise<Object>} Updated assignment data
 */
export const updateVehicleAssignment = async (assignmentId, updateData) => {
  if (!assignmentId) {
    throw parseApiError({
      response: {
        status: 400,
        data: { message: 'Assignment ID is required', error_code: ERROR_TYPES.VALIDATION },
      },
    });
  }

  return withRetry(
    async () => {
      const response = await httpClient.put(ASSIGNMENT_ENDPOINTS.update(assignmentId), updateData);
      return handleApiResponse(response);
    },
    { maxRetries: 1 }
  );
};

/**
 * Delete a vehicle assignment
 * @param {string} assignmentId - Assignment ID
 * @returns {Promise<Object>} Deletion confirmation
 */
export const deleteVehicleAssignment = async assignmentId => {
  if (!assignmentId) {
    throw parseApiError({
      response: {
        status: 400,
        data: { message: 'Assignment ID is required', error_code: ERROR_TYPES.VALIDATION },
      },
    });
  }

  return withRetry(
    async () => {
      const response = await httpClient.delete(ASSIGNMENT_ENDPOINTS.delete(assignmentId));
      return handleApiResponse(response);
    },
    { maxRetries: 1 }
  );
};

/**
 * Complete a vehicle assignment
 * @param {string} assignmentId - Assignment ID to complete
 * @param {Object} completionData - Completion data (notes, completion time, etc.)
 * @returns {Promise<Object>} Updated assignment data
 */
export const completeVehicleAssignment = async (assignmentId, completionData = {}) => {
  if (!assignmentId) {
    throw parseApiError({
      response: {
        status: 400,
        data: { message: 'Assignment ID is required', error_code: ERROR_TYPES.VALIDATION },
      },
    });
  }

  return withRetry(
    async () => {
      const endpoint = ASSIGNMENT_ENDPOINTS.complete(assignmentId);
      const response = await httpClient.put(endpoint, completionData);
      return handleApiResponse(response);
    },
    { maxRetries: 1 }
  );
};

/**
 * Cancel a vehicle assignment
 * @param {string} assignmentId - Assignment ID to cancel
 * @param {Object} cancellationData - Cancellation data (reason, cancellation time, etc.)
 * @returns {Promise<Object>} Updated assignment data
 */
export const cancelVehicleAssignment = async (assignmentId, cancellationData = {}) => {
  if (!assignmentId) {
    throw parseApiError({
      response: {
        status: 400,
        data: { message: 'Assignment ID is required', error_code: ERROR_TYPES.VALIDATION },
      },
    });
  }

  return withRetry(
    async () => {
      const endpoint = ASSIGNMENT_ENDPOINTS.cancel(assignmentId);
      const response = await httpClient.put(endpoint, cancellationData);
      return handleApiResponse(response);
    },
    { maxRetries: 1 }
  );
};

/**
 * Get assignment metrics and analytics
 * @param {boolean} useCache - Whether to use cached data
 * @returns {Promise<Object>} Assignment metrics data
 */
export const getAssignmentAnalytics = async (useCache = true) => {
  return withRetry(async () => {
    const url = `${ASSIGNMENT_ENDPOINTS.metrics}?use_cache=${useCache}`;
    const response = await httpClient.get(url);
    return handleApiResponse(response);
  });
};

/**
 * Get assignments for a specific vehicle
 * @param {string} vehicleId - Vehicle ID
 * @param {string} status - Optional status filter
 * @returns {Promise<Object>} Vehicle assignments
 */
export const getVehicleAssignmentsByVehicle = async (vehicleId, status = null) => {
  if (!vehicleId) {
    throw parseApiError({
      response: {
        status: 400,
        data: { message: 'Vehicle ID is required', error_code: ERROR_TYPES.VALIDATION },
      },
    });
  }

  const params = { vehicle_id: vehicleId };
  if (status) params.status = status;

  return await getVehicleAssignments(params);
};

/**
 * Get assignments for a specific driver
 * @param {string} driverId - Driver ID
 * @param {string} status - Optional status filter
 * @returns {Promise<Object>} Driver assignments
 */
export const getVehicleAssignmentsByDriver = async (driverId, status = null) => {
  if (!driverId) {
    throw parseApiError({
      response: {
        status: 400,
        data: { message: 'Driver ID is required', error_code: ERROR_TYPES.VALIDATION },
      },
    });
  }

  const params = { driver_id: driverId };
  if (status) params.status = status;

  return await getVehicleAssignments(params);
};

File: Core/routes/api/assignments.py
"""
Vehicle Assignment Management Routes
Handles all assignment-related operations through service proxy
"""

from fastapi import APIRouter, HTTPException, Depends, Request, Query
from fastapi.security import HTTPAuthorizationCredentials
from typing import Dict, Any, Optional
import logging

from .base import security, handle_service_request, validate_required_fields
from utils.exceptions import ValidationError

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/api", tags=["Vehicle Assignments"])

@router.get("/assignments")
async def get_assignments(
    request: Request,
    vehicle_id: Optional[str] = Query(None, description="Filter by vehicle ID"),
    driver_id: Optional[str] = Query(None, description="Filter by driver ID"),
    status: Optional[str] = Query(None, description="Filter by status"),
    credentials: HTTPAuthorizationCredentials = Depends(security)
):
    """Get all assignments via Management service"""
    logger.info(f"Received get_assignments request with params: {dict(request.query_params)}")
    
    response = await handle_service_request(
        endpoint="/api/assignments",
        method="GET",
        data=dict(request.query_params),
        credentials=credentials,
        auth_endpoint="/api/assignments"
    )
    
    return response

@router.post("/assignments")
async def create_assignment(
    assignment_data: Dict[str, Any],
    credentials: HTTPAuthorizationCredentials = Depends(security)
):
    """Create assignment via Management service"""
    # Validate input data
    required_fields = ["vehicle_id", "driver_id"]
    validate_required_fields(assignment_data, required_fields)
    
    # Add default values for backend compatibility
    if "status" not in assignment_data:
        assignment_data["status"] = "active"
    
    response = await handle_service_request(
        endpoint="/api/assignments",
        method="POST",
        data=assignment_data,
        credentials=credentials,
        auth_endpoint="/api/assignments"
    )
    
    return response

@router.get("/assignments/{assignment_id}")
async def get_assignment(
    assignment_id: str,
    credentials: HTTPAuthorizationCredentials = Depends(security)
):
    """Get specific assignment via Management service"""
    response = await handle_service_request(
        endpoint=f"/api/assignments/{assignment_id}",
        method="GET",
        data={"assignment_id": assignment_id},
        credentials=credentials,
        auth_endpoint="/api/assignments"
    )
    
    return response

@router.put("/assignments/{assignment_id}")
async def update_assignment(
    assignment_id: str,
    assignment_data: Dict[str, Any],
    credentials: HTTPAuthorizationCredentials = Depends(security)
):
    """Update assignment via Management service"""
    logger.info(f"Updating assignment {assignment_id} with data: {assignment_data}")
    
    response = await handle_service_request(
        endpoint=f"/api/assignments/{assignment_id}",
        method="PUT",
        data=assignment_data,
        credentials=credentials,
        auth_endpoint="/api/assignments"
    )
    
    logger.info(f"Assignment {assignment_id} updated successfully")
    return response

@router.delete("/assignments/{assignment_id}")
async def delete_assignment(
    assignment_id: str,
    credentials: HTTPAuthorizationCredentials = Depends(security)
):
    """Delete assignment via Management service"""
    logger.info(f"Deleting assignment {assignment_id}")
    
    response = await handle_service_request(
        endpoint=f"/api/assignments/{assignment_id}",
        method="DELETE",
        data={"assignment_id": assignment_id},
        credentials=credentials,
        auth_endpoint="/api/assignments"
    )
    
    logger.info(f"Assignment {assignment_id} deleted successfully")
    return response

@router.put("/assignments/{assignment_id}/complete")
async def complete_assignment(
    assignment_id: str,
    completion_data: Dict[str, Any],
    credentials: HTTPAuthorizationCredentials = Depends(security)
):
    """Complete assignment via Management service"""
    logger.info(f"Completing assignment {assignment_id}")
    
    response = await handle_service_request(
        endpoint=f"/api/assignments/{assignment_id}/complete",
        method="PUT",
        data=completion_data,
        credentials=credentials,
        auth_endpoint="/api/assignments"
    )
    
    logger.info(f"Assignment {assignment_id} completed successfully")
    return response

@router.put("/assignments/{assignment_id}/cancel")
async def cancel_assignment(
    assignment_id: str,
    cancellation_data: Dict[str, Any],
    credentials: HTTPAuthorizationCredentials = Depends(security)
):
    """Cancel assignment via Management service"""
    logger.info(f"Cancelling assignment {assignment_id}")
    
    response = await handle_service_request(
        endpoint=f"/api/assignments/{assignment_id}/cancel",
        method="PUT",
        data=cancellation_data,
        credentials=credentials,
        auth_endpoint="/api/assignments"
    )
    
    logger.info(f"Assignment {assignment_id} cancelled successfully")
    return response
